implemented prometheus-metrics as flask middleware at /metrics

tobischenk opened the pull request:
This Pull request is thought to add prometheus metrics to pyca. The metrics are retrievable at '/metrics'. The metrics include prometheus-client default metrics and two custom collectors to return metrics about the status of pyca services and info about upcoming and recorded events.

Changes:
- setup.py, requirements.txt: Added prometheus-client library
- pyca/ui/__init__.py: changed to include prometheus metrics as flask middleware at endpoint /metrics
- pyca/ui/process_status_collector.py: Custom Collector for pyca service status
- pyca/ui/recordings_collector.py: Custom Collector for recorded and upcoming events

lkiesow commented on pyca/ui/process_status_collector.py:
```diff
@@ -0,0 +1,51 @@
+
+    @staticmethod
+    def get_state_dict(state: str):
+        return {
+            'stopped': 'stopped' == state,
+            'idle': 'idle' == state,
+            'busy': 'busy' == state,
+        }
```
Converting this to strings in between is unnecessary and with the change proposed for recordings, you can just do:

```suggestion
        return { ServiceStatus.str(s):s == state for s in ServiceStatus.values() }
```
and then something like:

```py
self.get_state_dict(get_service_status(Service.AGENTSTATE))
```

lkiesow commented on pyca/ui/process_status_collector.py:
```diff
@@ -0,0 +1,51 @@
+            labels=['service', 'state']
+        )
+
+        service_states.add_metric(
+            ['capture'],
+            self.get_state_dict(ServiceStatus.str(get_service_status(Service.CAPTURE)))
+        )
```
You can do somthing like:

```suggestion
        for service in Service.values():
            service_states.add_metric(
                [Service.str(service)],
                self.get_state_dict(get_service_status(service))
            )
```

lkiesow reviewed (changes requested):
Hi @tobischenk,
thanks for the patch.

As you have probably seen, the tests fail on this pull request. The CodeQL seems to have started complaining about old code. Feel free to ignore that. The other tests should succees though. Can you fix your code accordingly?

I've also added some comments related to the code itself.
Let me know if you think any of the comments make no sense.

Finally, it would be great to add tests for the new code.

lkiesow commented:
CodeQL check is fixed.

tobischenk commented:
I fixed the linting errors regarding unused import statements in https://github.com/opencast/pyCA/pull/336/commits/a28bdb4d1572faea21e929d007e7b53774b063c3

In https://github.com/opencast/pyCA/pull/336/commits/9d7f4269ec72ce6925e189792e3891f81be1b677 i added the values method to pyca/db.py Constant in order to use it in the custom collectors.

I adjusted the custom collectors in https://github.com/opencast/pyCA/pull/336/commits/6527e5c48df07be488b979f3b027b65b5b18aa55 and fixed linting errors for too long lines, but I am unsure if understood the calculation in  https://github.com/opencast/pyCA/pull/336#discussion_r684403415 correctly.

lkiesow commented:
> but I am unsure if understood the calculation in #336 (comment) correctly.

You mean the calculation of upcoming events?

tobischenk commented:
> > but I am unsure if understood the calculation in #336 (comment) correctly.
> 
> You mean the calculation of upcoming events?

Yes exactly. I am not sure if the implementation in https://github.com/tobischenk/pyCA/blob/6527e5c48df07be488b979f3b027b65b5b18aa55/pyca/ui/recordings_collector.py#L32-L41 returns the correct count.

Moreover I removed unused dependencies, which caused the test to fail in ce961c370026fc2777aeaa642a2b626f32c78d72.

Maybe you can run the test again so i can check is something is still failing?

tobischenk commented:
Since the test for the Code passed, I will try to add tests in the following week. I will commit the changes and write a new comment once its done.

lkiesow commented:
I'll also try to write up a few details about the upcoming events table. Ping me if I forget :-)

lkiesow commented:
Here comes a short explanation of how the two recording tables work and why there are two in the first place.
Let's start with what the tables are doing:

- `upcoming_event` acts like a cache for whatever events Opencast lists for pyCA. It is more or less a semi-persistent representation of Opencast's iCal feed for the capture agent. The table will be truncated and refilled with each successful update.
- `recorded_event` holds the events pyCA has accepted and started to process. It is also used to keep track of the recording's state. pyCA may modify events in here, store additional data, …

So, `recorded_event` is more or less a classic database table, and you can interact with it the way you would expect. The `upcoming_event` may be a bit more confusing, however.

First, this table exists just as a cache. The reason for this is that we want to compensate for common infrastructure failures as good as we can. These include network failures or power failures. The way it is build, pyCA can loose network and power but as soon as it starts up again, it can work independently and capture events to the best of his knowledge, even if the network doesn't come up again. This is something I've actually seen in real life. Networks aren't always as resilient as one would think and if there is a larger problem, one lecture hall's network may also not be the top priority.

This means whatever Opencast sends us is in `recorded_event` and that's all upcoming recordings, including those which should already have started. The reasoning for Opencast including those is similar: If a capture agent connects late or gets the update late, it is better to start a bit late, than to not start at all.

What now happens is that pyCA picks a recording, starts processing and copies the recording to `recorded_event`. But the same recording is still in `upcoming_event`. Even if we would remove the event from that table when we start the next recording, it would be back again with the next update we get from Opencast. Besides, we use this for trying to restart events which capture command failed at some point.

What this boils down to is that there may be events in `upcoming_event` which are also already tracked in `recorded_event`. To count the events which aren't in another state and are indeed still upcoming, we would thus need to count the events which are not present in `recorded_event` (e.g. check for each event if it is in that table and exclude the event if it is). But we can make this a bit easier by using some internal knowledge: In production, only the newest event in `recorded_event` should potentially be present in `upcoming_event`. So, you could do something like (pseudo code):

```
latest_recorded_id = select id from recorded_event order by date desc limit 0, 1;
upcoming = select count(*) from upcoming_event where id != latest_recorded_id;
```

Now, if you are strict about it, you could potentially construct cases where this is not true by deleting and thus creating some overlapping last minute events in Opencast but given that this is about metrics, I think we can safely ignore that.

lkiesow commented:
…actually, thinking about this for a moment, I think we can even get the edge cases relatively easy without comparing everything by doing something like:

```
# Get list of events which are potentially still in upcoming
recorded_ids = select uid from recorded_event where end > timestamp();
# Count all events which aren't already in recorded_event
upcoming = select count(*) from upcoming_event where not uid in recorded_ids;
```

tobischenk commented:
Hello Ikiesow,
I changed the calculation of upcoming events. Moreover I tried to write unit test, but I am unsure if they cover all the lines, since its the first time im writing unit tests.

If anything is not according to standards and can be improved, please let me know.

Thanks in advance.

tobischenk commented:
@lkiesow can you run the tests again?

Regards tobischenk

lkiesow commented:
Sorry for the late reply. Didn't manage to get this merged before my vacation after all.
This looks good → merging.
This also means that all tests will now run automatically for you on new pull requests ;-)

Thanks again for the patch.
I think it is quite useful.
